src/actions/mod.rs

```diff
@@ -33,7 +34,63 @@ use std::time::Duration;
 
 use self::compiler_message_parsing::{FileDiagnostic, ParseError};
 
-type BuildResults = HashMap<PathBuf, Vec<Diagnostic>>;
+type BuildResults = HashMap<PathBuf, Vec<RustDiagnostic>>;
+
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+pub struct PublishRustDiagnosticsParams {
+    /// The URI for which diagnostic information is reported.
+    #[serde(deserialize_with = "url_serde::deserialize", serialize_with = "url_serde::serialize")]
+    pub uri: Url,
+
+    /// An array of diagnostic information items.
+    pub diagnostics: Vec<RustDiagnostic>,
+}
+
+impl PublishRustDiagnosticsParams {
+    pub fn new(uri: Url, diagnostics: Vec<RustDiagnostic>) -> PublishRustDiagnosticsParams {
+        PublishRustDiagnosticsParams {
+            uri: uri,
+            diagnostics: diagnostics,
+        }
+    }
+}
+
+/// A range in a text document expressed as (zero-based) start and end positions.
+/// A range is comparable to a selection in an editor. Therefore the end position is exclusive.
+#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
+pub struct LabelledRange {
+    /// The range's start position.
+    pub start: Position,
+    /// The range's end position.
+    pub end: Position,
+    /// The optional LabelledRange.
+    pub label: Option<String>,
+}
+
+/// Represents a diagnostic, such as a compiler error or warning.
+/// Diagnostic objects are only valid in the scope of a resource.
+#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
+pub struct RustDiagnostic {
+    /// The primary range at which the message applies.
+    pub range: LabelledRange,
+
+    /// The secondary ranges that apply to the message
+    pub secondary: Vec<LabelledRange>,
```

Comment:
I find it a bit odd to have a field called 'secondary' - it is a list, not a single value and there is no 'primary' filed. I would prefer `children`, but happy to have something less generic too.